### ISSUE TITLE
Request for clarification of implementation details of  `p_sample`

### ISSUE THREAD

**Issue by lassepe** [LiorSinai/DenoisingDiffusion.jl#7]
Hi @LiorSinai,

Thank you for sharing your implementation of denoising diffusion probabilistic models. I'm excited to see this being available in Julia!

This isn't an issue but rather a request for clarification: when I went over your implementation of `GaussianDiffusion,` I was surprised to see that your sampling procedure in the reverse process is different from the one described in the original paper by Jonathan Ho et al.

Specifically, in your implementation, you compute `x_start` (x0 in Ho's paper) from the output of the denoising network and then use that to sample from `q_{posterior}(x_{t-1} | x_t, x_0) = q(x_t | x_{t-1}, x_0) q(x_{t-1} | x_0) / q(x_t | x_0)`

https://github.com/LiorSinai/DenoisingDiffusion.jl/blob/cedf76064265c624921290a36983f2273de2973b/src/GaussianDiffusion.jl#L183-L197

I cannot find a reference to this sampling strategy in the literature, so I was wondering if you could share the motivation (or a reference) for this parameterization. In particular, the original paper includes the comment:

> There is also the possibility of predicting x0, but we found this to lead to worse sample quality early in our experiments.

**Comment by LiorSinai** [comment]
I find your question curious and I do not see how my sampling method differs. 

The aim of this repository is to implement the original paper with few of the later proposed modifications (latent diffusion, transformer based models, cold diffusion etc.).

The original paper uses equation 4 for `x0` and substitutes it into equation 7 to obtain a simplified formula for equation 11. 

However this repository implements equation 7 directly in the following: https://github.com/LiorSinai/DenoisingDiffusion.jl/blob/cedf76064265c624921290a36983f2273de2973b/src/GaussianDiffusion.jl#L152-L158

The reason for not using equation 11 is that it is useful to keep track of `x0` as well, and it is interesting to see how it evolves throughout the process.

Here `xstart` comes from: https://github.com/LiorSinai/DenoisingDiffusion.jl/blob/cedf76064265c624921290a36983f2273de2973b/src/GaussianDiffusion.jl#L171-L175

which shows that indeed the `noise` ε is being predicted by the model, and `xstart` is calculated using the reparameterized form of equation 4:
https://github.com/LiorSinai/DenoisingDiffusion.jl/blob/cedf76064265c624921290a36983f2273de2973b/src/GaussianDiffusion.jl#L165-L169

The name `predict_start_from_noise` from noise might be a misleading, because it is directly calculated and not predicted by the model. The name comes this [denoising-diffusion-pytorch](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/a251d9ab2547b257a60c9d9fc7380d2a10239194/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L625) repository that I based this on. Note that unlike this repository, that one gives you a choice for what to predict with the model. However as the comment you quoted says, it is best to predict the noise, so that is the only option here. 

I recommend checking out this section in my [blog post](https://liorsinai.github.io/coding/2022/12/03/denoising-diffusion-1-spiral.html#reverse-process) for a full explanation.

**Comment by lassepe** [comment]
Thank you for the swift response. Your clarification was helpful and I think I see now what's going on.

I'm adding my conclusion below for future reference for others and also for you to challenge it in case you think my understanding is incorrect:

The thing that threw me off is the fact that $q$ in equation (4) describes the *forward* process and $q(x_{t-1}\mid x_t, x_0)$ is the true posterior distribution over $x_{t-1}$ given $x_t$ and *known* $x_0$. In the context of training, $x_0$ is known and there's no reason to express it in terms of $\epsilon$ (and indeed, during training we evaluate $q(x_{t-1}\mid x_t, x_0)$ with the "ground-truth" $x_0$; when I opened this issue, I didn't see that you are actually not using `p_sample` in the loss but only at test time). 

Of course, at test time, we do not know $x_0$ and we have to instead sample $x_{t-1}$ only knowing $x_t$ and the predicted noise $\epsilon_\theta$. So we use $\epsilon_\theta$ to construct an estimate of $\tilde{\mu}_t$ in equation (7) based on the previously sampled $x_t$ and our maximum-likelihood-estimate  $\hat{x}_0 = \frac{1}{\sqrt{\bar{\alpha}_t}} (x_t - \sqrt{1 - \bar{\alpha}_t} \epsilon\theta )$

**Comment by LiorSinai** [comment]
Yes indeed. This is what I wrote in my blog post, which is an informal version of what you have stated:

> [G]iven the current time step and the start image, we can with a high probability deduce what the previous time step was based on the known probability distributions. Of course this statement is nonsensical in that we don’t have the start image - if we did we wouldn’t need to go through this process. However what we can do is use a model to predict a start image and refine it at every time step. The analogy I used earlier therefore needs correcting: while we can’t jump straight from noise to a good starting image, we can estimate a bad start image and refine it.
